### PR TITLE
refactor: Gas Optimizations (C4 and other) (SC-2038)

### PR DESCRIPTION
**Pull request by deluca-mike**
# Description

Various gas optimizations, including those specified here: https://app.clubhouse.io/maplefinance/story/2038/maple-core-c4-gas-optimizations

# Integrations Checklist

- [X] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [X] Has all documentation been updated?

# Changelog

chore: Gas Optimization (https://github.com/code-423n4/2021-04-maple-findings/issues/11)
- Unused definition of enum (https://github.com/code-423n4/2021-04-maple-findings/issues/17)
- Using precalculated value in the Pool & Loan contracts (https://github.com/code-423n4/2021-04-maple-findings/issues/75)
- Unused variable (https://github.com/code-423n4/2021-04-maple-findings/issues/78)

refactor: More gas optimizations
- Use local variables to reduce storage access
- Use return variables when legible and possible
- if (exitCase) exit, so that large code can be done unindented, and smaller jumpdest
- --x and ++x are sometimes better, since there is no intermediate variable on the stack

### CODE REVIEW

File: contracts/FundingLockerFactory.sol [hunk 1]
```diff
@@ -18,15 +18,14 @@ contract FundingLockerFactory {
...
-        @return Address of the instantiated funding locker
+        @param liquidityAsset The asset this funding locker will escrow.
+        @return fundingLocker Address of the instantiated funding locker.
     */
-    function newLocker(address liquidityAsset) public returns (address) {
-        address fundingLocker   = address(new FundingLocker(liquidityAsset, msg.sender));
+    function newLocker(address liquidityAsset) external returns (address fundingLocker) {
+        fundingLocker   = address(new FundingLocker(liquidityAsset, msg.sender));
         owner[fundingLocker]    = msg.sender;
         isLocker[fundingLocker] = true;
```
Comment (lucas-manuel) [inline contracts/FundingLockerFactory.sol:27]: `=` alignment

File: contracts/LiquidityLockerFactory.sol [hunk 2]
```diff
@@ -16,15 +16,14 @@ contract LiquidityLockerFactory {
...
-        @return Address of the instantiated liquidity locker
+        @param liquidityAsset   The asset this liquidity locker will escrow.
+        @return liquidityLocker Address of the instantiated liquidity locker.
     */
-    function newLocker(address liquidityAsset) external returns (address) {
-        address liquidityLocker   = address(new LiquidityLocker(liquidityAsset, msg.sender));
+    function newLocker(address liquidityAsset) external returns (address liquidityLocker) {
+        liquidityLocker   = address(new LiquidityLocker(liquidityAsset, msg.sender));
         owner[liquidityLocker]    = msg.sender;
         isLocker[liquidityLocker] = true;
```
Comment (lucas-manuel) [inline contracts/LiquidityLockerFactory.sol:26]: `=` alignment

File: contracts/Loan.sol [hunk 3]
```diff
@@ -135,8 +135,8 @@ contract Loan is FDT, Pausable {
         address[3] memory calcs
     )
         FDT(
-            string(abi.encodePacked("Maple Loan Token")),
-            string(abi.encodePacked("MPL-LOAN")),
+            "Maple Loan Token",
+            "MPL-LOAN",
             _liquidityAsset
         )
```
Comment (lucas-manuel) [inline contracts/Loan.sol:141]: Lets just make this a one liner now actually

File: contracts/MplRewards.sol [hunk 4]
```diff
@@ -72,11 +74,11 @@ contract MplRewards is Ownable {
...
     function rewardPerToken() public view returns (uint256) {
-        if (_totalSupply == 0) return rewardPerTokenStored;
-        return
-            rewardPerTokenStored.add(
-                lastTimeRewardApplicable().sub(lastUpdateTime).mul(rewardRate).mul(1e18).div(_totalSupply)
-            );
+        return _totalSupply == 0
+            ? rewardPerTokenStored
+            : rewardPerTokenStored.add(
+                   lastTimeRewardApplicable().sub(lastUpdateTime).mul(rewardRate).mul(1e18).div(_totalSupply)
```
Comment (lucas-manuel) [inline contracts/MplRewards.sol:80]: Should be four spaces here

File: contracts/PremiumCalc.sol [hunk 5]
```diff
@@ -21,15 +21,15 @@ contract PremiumCalc {
...
+        @return principalOwed Principal
+        @return interest      Interest
     */
-    function getPremiumPayment(address _loan) public view returns(uint256, uint256, uint256) {
-        ILoan   loan          = ILoan(_loan);
-        uint256 principalOwed = loan.principalOwed();
-        uint256 interest      = principalOwed.mul(premiumFee).div(10_000);
-        return (interest.add(principalOwed), principalOwed, interest);
+    function getPremiumPayment(address _loan) external view returns(uint256 total, uint256 principalOwed, uint256 interest) {
+        ILoan   loan  = ILoan(_loan);
```
Comment (lucas-manuel) [inline contracts/PremiumCalc.sol:30]: `ILoan loan` and then align `=`

File: contracts/StakeLocker.sol [hunk 6]
```diff
@@ -62,9 +62,11 @@ contract StakeLocker is StakeLockerFDT, Pausable {
                2. The Pool is in Initialized or Deactivated state.
     */
     modifier canUnstake() {
+        IPool poolConrtact = IPool(pool);
```
Comment (lucas-manuel) [inline contracts/StakeLocker.sol:65]: Lets just use `_pool` here, thats the pattern we've been using with the `IMapleGlobals` interface

File: contracts/StakeLocker.sol [hunk 7]
```diff
@@ -178,14 +182,15 @@ contract StakeLocker is StakeLockerFDT, Pausable {
...
-        }
+        uint256 balance = balanceOf(who);
+
+        // stakeDate + (now - stakeDate) * (amt / (balance + amt))
+        // NOTE: prevDate = 0 implies balance = 0, and equation reduces to now
+        uint256 newDate = balance + amt > 0
+            ? prevDate.add(block.timestamp.sub(prevDate).mul(amt).div(balance + amt))
+            : prevDate;
+
+        stakeDate[who] = newDate;
```
Comment (lucas-manuel) [inline contracts/StakeLocker.sol:193]: Wasn't this change in the other PR? Guess it still needs to get merged into this.

File: contracts/StakeLocker.sol [hunk 8]
```diff
@@ -155,18 +157,20 @@ contract StakeLocker is StakeLockerFDT, Pausable {
         @param amt Amount of stakeAsset (BPTs) to deposit
     */
     function stake(uint256 amt) whenNotPaused external {
+        address msgSender = msg.sender;
+
         _whenProtocolNotPaused();
-        _isAllowed(msg.sender);
+        _isAllowed(msgSender);
```
Comment (lucas-manuel) [inline contracts/StakeLocker.sol:163]: Lets revert all instances of `msgSender` and `blockTimestamp`

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Initial review

File: contracts/StakeLocker.sol (hunk 7 above)
Comment (deluca-mike) [reply contracts/StakeLocker.sol:191]: Yup. Once I rebase on the other, it will be gone. I am not a git pro, so I don't really know how to avoid this.

File: contracts/MplRewards.sol [hunk 9]
```diff
@@ -139,20 +141,21 @@ contract MplRewards is Ownable {
...
-            uint256 remaining = periodFinish.sub(block.timestamp);
-            uint256 leftover  = remaining.mul(rewardRate);
-            rewardRate        = reward.add(leftover).div(rewardsDuration);
-        }
+
+        uint256 _rewardRate = block.timestamp >= periodFinish
+            ? reward.div(rewardsDuration)
+            : reward.add(
+                  periodFinish.sub(block.timestamp).mul(rewardRate)
+              ).div(rewardsDuration);
```
Comment (lucas-manuel) [inline contracts/MplRewards.sol:149]: @satyamakgec @michaelelliot This LGTM but I'd like more eyes on it

File: contracts/token/ExtendedFDT.sol [hunk 10]
```diff
@@ -53,9 +53,11 @@ abstract contract ExtendedFDT is BasicFDT {
 
         if (value == 0) return;
 
-        lossesPerShare = lossesPerShare.add(value.mul(pointsMultiplier) / totalSupply());
+        uint256 perShareLosses = lossesPerShare.add(value.mul(pointsMultiplier) / totalSupply());
```
Comment (lucas-manuel) [inline contracts/token/ExtendedFDT.sol:56]: Lets change this to `lossesPerShare`

File: contracts/token/ExtendedFDT.sol [hunk 11]
```diff
@@ -66,9 +68,10 @@ abstract contract ExtendedFDT is BasicFDT {
     function _prepareLossesWithdraw() internal returns (uint256 recognizableDividend) {
         recognizableDividend = recognizableLossesOf(msg.sender);
 
-        recognizedLosses[msg.sender] = recognizedLosses[msg.sender].add(recognizableDividend);
+        uint256 _recognizedLosses = recognizedLosses[msg.sender].add(recognizableDividend);
+        recognizedLosses[msg.sender]   = _recognizedLosses;
```
Comment (lucas-manuel) [inline contracts/token/ExtendedFDT.sol:72]: Align `=`

**Review by lucas-manuel — APPROVED** [review]
LGTM, some small changes. Would like a review from @satyamakgec and/or @michaelelliot

File: contracts/Loan.sol [hunk 12]
```diff
@@ -246,7 +239,7 @@ contract Loan is FDT, Pausable {
         _whenProtocolNotPaused();
         _isValidState(State.Active);
         (uint256 total, uint256 principal, uint256 interest,, bool paymentLate) = getNextPayment();
-        paymentsRemaining--;
+        --paymentsRemaining;
```
Comment (satyamakgec) [inline contracts/Loan.sol:242]: If I am not wrong this will update the `paymentRemaining` at last, I am making a guess from the C knowledge I have. And what is the rationale of this change?

File: contracts/LoanFactory.sol [hunk 13]
```diff
@@ -107,7 +107,7 @@ contract LoanFactory is Pausable {
         loanAddress         = address(loan);
         loans[loansCreated] = loanAddress;
         isLoan[loanAddress] = true;
-        loansCreated++;
+        ++loansCreated;
```
Comment (satyamakgec) [inline contracts/LoanFactory.sol:110]: ditto

File: contracts/MplRewards.sol [hunk 14]
```diff
@@ -121,11 +123,11 @@ contract MplRewards is Ownable {
         _updateReward(msg.sender);
         uint256 reward = rewards[msg.sender];
 
-        if (reward > 0) {
-            rewards[msg.sender] = 0;
-            rewardsToken.safeTransfer(msg.sender, reward);
-            emit RewardPaid(msg.sender, reward);
-        }
+        if (reward == 0) return;
```
Comment (satyamakgec) [inline contracts/MplRewards.sol:126]: could use proper casting, `uint256(0)`

File: contracts/MplRewards.sol [hunk 15]
```diff
@@ -121,11 +123,11 @@ contract MplRewards is Ownable {
...
+
+        rewards[msg.sender] = 0;
```
Comment (satyamakgec) [inline contracts/MplRewards.sol:128]: uint256(0)

File: contracts/PoolFactory.sol [hunk 16]
```diff
@@ -92,12 +93,13 @@ contract PoolFactory is Pausable {
                 symbol
             );
 
-        pools[poolsCreated]   = address(pool);
-        isPool[address(pool)] = true;
-        poolsCreated++;
+        poolAddress         = address(pool);
+        pools[poolsCreated] = poolAddress;
+        isPool[poolAddress] = true;
+        ++poolsCreated;
```
Comment (satyamakgec) [inline contracts/PoolFactory.sol:99]: ditto

File: contracts/PremiumCalc.sol [hunk 17]
```diff
@@ -21,15 +21,15 @@ contract PremiumCalc {
...
+        @return interest      Interest
     */
-    function getPremiumPayment(address _loan) public view returns(uint256, uint256, uint256) {
-        ILoan   loan          = ILoan(_loan);
-        uint256 principalOwed = loan.principalOwed();
-        uint256 interest      = principalOwed.mul(premiumFee).div(10_000);
-        return (interest.add(principalOwed), principalOwed, interest);
+    function getPremiumPayment(address _loan) external view returns(uint256 total, uint256 principalOwed, uint256 interest) {
+        ILoan loan    = ILoan(_loan);
+        principalOwed = loan.principalOwed();
```
Comment (satyamakgec) [inline contracts/PremiumCalc.sol:31]: We can directly use `ILoan(_loan)` as it not again used

File: contracts/Loan.sol (hunk 12 above)
Comment (deluca-mike) [reply contracts/Loan.sol:242]: `++x` and `--x` are immediate, where as `x++` or `x--` keeps x on the stack to be used for ops (nothing in this case) and then increments or decrements. I have done a good amount of testing with this and, because solc is perfect, I have noticed its always best to `--x` or `++x` if you don't need the post-decrement/increments effects.

**Review by satyamakgec — APPROVED** [review]

**Review by lucas-manuel — APPROVED** [review]
